### PR TITLE
Daemon/Stages/Patterns: add inlay type hints

### PR DESCRIPTION
**Pull request by DedSec256**
Currently supported: 

- Top/Local bindings & type members parameters/return types
- Foreach patterns
- Lambda parameters

Settings:

{3B3F5DCE-A9ED-4188-ABB6-25151AF82DD2}

Small demo:

example

Examples:

{FB04E510-50BA-4A42-AFCE-688ED3D851F0}

{555B8269-4C64-4012-9DFC-6CBCD12EA0F8}

{2A8F2820-1A41-4007-8091-16040B79378C}

### CODE REVIEW

File: ReSharper.FSharp/src/FSharp/FSharp.Common/src/Settings/FSharpOptions.fs [hunk 1]
```diff
@@ -173,13 +177,23 @@ module FSharpTypeHintOptions =
...
       [<SettingsEntry(true, FSharpTypeHintOptions.hideSameLinePipe); DefaultValue>]
-      mutable HideSameLine: bool }
+      mutable HideSameLine: bool
+
+      [<SettingsEntry(PushToHintMode.Default,
+                      DescriptionResourceType = typeof<Strings>,
+                      DescriptionResourceName = nameof(Strings.FSharpTypeHints_TopLevelMembers_Description))>]
+      mutable ShowTypeHintsForTopLevelMembers: PushToHintMode
+
+      [<SettingsEntry(PushToHintMode.PushToShowHints,
```
Comment (DedSec256) [inline ReSharper.FSharp/src/FSharp/FSharp.Common/src/Settings/FSharpOptions.fs:193]: Perhaps we can make a different default here, but this one seemed most appropriate to me

File: ReSharper.FSharp/src/FSharp/FSharp.Common/src/Settings/FSharpOptions.fs [hunk 2]
```diff
@@ -173,13 +177,23 @@ module FSharpTypeHintOptions =
...
+                      DescriptionResourceType = typeof<Strings>,
+                      DescriptionResourceName = nameof(Strings.FSharpTypeHints_LocalBindings_Description))>]
+      mutable ShowTypeHintsForLocalBindings: PushToHintMode }
```
Comment (DedSec256) [inline ReSharper.FSharp/src/FSharp/FSharp.Common/src/Settings/FSharpOptions.fs:196]: Perhaps we can make a different default here, but this one seemed most appropriate to me

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Options/FSharpTypeHintOptionsPage.fs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+              "Type Hints",
+              null,
+              ParentId = nameof(FSharpInlayHintsPage),
+              NestingType = OptionPageNestingType.Child,
+              Sequence = 1.,
+              NameResourceType = typeof<Strings>,
+              NameResourceName = nameof(Strings.FSharpTypeHints_OptionsPage_Title))>]
+type FSharpTypeHintsOptionsPage(lifetime: Lifetime, optionsPageContext: OptionsPageContext,
+                                optionsSettingsSmartContext: OptionsSettingsSmartContext) as this =
+    inherit InlayHintsOptionPageBase(lifetime, optionsPageContext, optionsSettingsSmartContext, optionsPageContext.IsReSharper)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Options/FSharpTypeHintOptionsPage.fs:33]: Can we move the `IsReSharper` check inside the base constructor? Though, it needs to be done outside of this PR.

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+    let isTopLevelMember (node: ITreeNode) =
+        match node with
+        | :? ITopBinding
+        | :? IMemberDeclaration
+        | :? IConstructorDeclaration -> true
+        | _ -> false
+
+    override x.VisitNode(node, context) =
+        if settings.LocalBindings = PushToHintMode.Never && isTopLevelMember node then () else
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:50]: Why does it check the settings for local bindings but checks whether a node is top level?

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+
+        for child in node.Children() do
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:52]: Does it iterate through all the children when the current mode should only show the hints for the top level nodes? Or do we need to create the hints anyway?

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+        let symbolUse =
+            match decl with
+            | :? IBinding as binding ->
+                let refPat = binding.HeadPattern.As<IReferencePat>()
+                if isNull refPat then Unchecked.defaultof<_> else
+                refPat.GetFcsSymbolUse()
+            | :? IMemberDeclaration as memberDeclaration ->
+                memberDeclaration.GetFcsSymbolUse()
+            | _ -> Unchecked.defaultof<_>
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:136]: Would getting it via `IFSharpDeclaration` work better?

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+        if isNull symbolUse then ValueNone else
+        let symbol = symbolUse.Symbol.As<FSharpMemberOrFunctionOrValue>()
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:139]: Could you please add empty lines before such `else` branches?

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+        | :? IAutoPropertyDeclaration as decl ->
+            let symbolUse = decl.GetFcsSymbolUse()
+            if isNull symbolUse then ValueNone else
+            let symbol = symbolUse.Symbol.As<FSharpMemberOrFunctionOrValue>()
+            if isNull symbol then ValueNone else
+
+            let fcsType = symbol.ReturnParameter.Type
+            let range = decl.NameIdentifier.GetNavigationRange().EndOffsetRange()
+            let displayContext = symbolUse.DisplayContext.WithShortTypeNames(true)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:186]: Parts of this code seem to be repeated here and there. Could you try to extract it and reuse?

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+        adornNodes localNodes settings.LocalBindings
+
+        highlightingConsumer.CollectHighlightings()
+
+    override x.Execute(committer) =
+        let consumer = { TopLevelNodes = List(); LocalNodes = List() }
+        fsFile.Accept(MembersVisitor(settings), consumer)
+
+        let topLevelNodes = consumer.TopLevelNodes |> Array.ofSeq
+        let localNodes = consumer.LocalNodes |> Array.ofSeq
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:214]: Could you explain why the arrays are needed, please?

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Services/src/Util/TypeAnnotationsUtil.fs [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    |> Seq.fold visitPattern acc
+
+let collectTypeHintAnchorsForBinding (binding: IBinding) =
+    let acc: ITreeNode list = []
+
+    if isNull binding || isNull binding.EqualsToken then acc else
+
+    let acc =
+        if isNotNull binding.ReturnTypeInfo then acc
+        elif binding.HasParameters then [(binding: ITreeNode)]
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi.Services/src/Util/TypeAnnotationsUtil.fs:51]: Are parens needed here?

File: ReSharper.FSharp/src/FSharp/FSharp.Psi/src/FSharp.psi [hunk 11]
```diff
@@ -1457,6 +1457,7 @@ autoPropertyDeclaration options { stubBase="FSharpProperTypeMemberDeclarationBas
   memberKeyword{MEMBER, MemberKeyword}
   VAL<VAL, ValKeyword>
   IDENTIFIER<IDENTIFIER, Identifier>
+  typeUsage<TYPE, TypeUsage>?
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi/src/FSharp.psi:1460]: Isn't the type usage separated from the name with a `:`?

File: ReSharper.FSharp/src/FSharp/FSharp.Psi/src/Impl/Tree/PrimaryConstructorDeclaration.cs [hunk 12]
```diff
@@ -13,6 +13,8 @@ public override TreeTextRange GetNameRange() =>
     protected override IDeclaredElement CreateDeclaredElement() =>
       new FSharpPrimaryConstructor(this);
 
+    public ITokenNode EqualsToken => null;
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi/src/Impl/Tree/PrimaryConstructorDeclaration.cs:16]: What do you think about moving the following here, from `collectTypeHintAnchorsForConstructor`?
```fsharp
let owner = FSharpTypeDeclarationNavigator.GetByPrimaryConstructorDeclaration(ctor)
isNotNull owner && isNotNull owner.EqualsToken
```

We could then have a doc comment saying something like ``Returns associated `=` token, possibly from containing type declaration``.

**Comment by auduchinok** [comment]
@DedSec256 Very nice work! What do you think about moving `git` authors to the right of declarations, so we could prevent the interlines in the editor?

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs (hunk 4 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:50]: The idea is not to visit children inside the top-level node if hints are not needed inside anyway.

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs (hunk 5 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:52]: If we want to show hints only for the top level, then the check above will not allow us to go inside the top binding. We create hints for children if the Always/PushToHint mode is enabled for local bindings.

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs (hunk 9 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:214]: Because of `Array.partition` function. Made by analogy with

https://github.com/JetBrains/resharper-fsharp/blob/f6475da23004a3eb8f2487e8a161e11ead450c26/ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PipeChainTypeHintStage.fs#L150-L164

Perhaps there is a more efficient way, but I didn’t immediately come up with a more elegant solution

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Services/src/Util/TypeAnnotationsUtil.fs (hunk 10 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp/FSharp.Psi.Services/src/Util/TypeAnnotationsUtil.fs:51]: Yeah 

{08508C3B-125C-4CC3-B53C-A395ADD78022}

{651D424A-581C-4B9F-8593-01C0C2B2737A}


it is also possible to write `binding :> ITreeNode`

File: ReSharper.FSharp/src/FSharp/FSharp.Psi/src/FSharp.psi (hunk 11 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp/FSharp.Psi/src/FSharp.psi:1460]: Sorry, this change should not be presented in **this PR**. I will remove it.

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs (hunk 6 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:136]: Oh, right now bindings are not an IFSharpDeclaration :(

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs (hunk 8 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:186]: It's good that you noticed. In general, this entire piece of code should be deleted, since we don't show hints for auto-properties (just like we didn't show Code Vision)

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Services/src/Util/TypeAnnotationsUtil.fs (hunk 10 above)
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp/FSharp.Psi.Services/src/Util/TypeAnnotationsUtil.fs:51]: Wow, OK. 😞

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/FSharp.Psi.Daemon.fsproj [hunk 13]
```diff
@@ -13,8 +13,6 @@
     <TargetFramework>net472</TargetFramework>
   </PropertyGroup>
   <ItemGroup>
-    <Compile Include="PsiFeaturesNamespaceZoneMarker.fs" />
-    <Compile Include="PsiNamespaceZoneMarker.fs" />
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/FSharp.Psi.Daemon.fsproj:17]: Are these changes needed in this PR?

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Options/FSharpTypeHintOptionsPage.fs (hunk 3 above)
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Options/FSharpTypeHintOptionsPage.fs:33]: @DedSec256 Please move the check inside the base constructor after this is merged to the monorepo.

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+        context.LocalNodes.AddRange(result)
+
+        x.VisitNode(forEachExpr, context)
+
+type private PatternsHighlightingProcess(logger: ILogger, fsFile, settingsStore: IContextBoundSettingsStore,
+                                           daemonProcess: IDaemonProcess, settings) =
+    inherit FSharpDaemonStageProcessBase(fsFile, daemonProcess)
+
+    let createTypeHintHighlighting (fcsType: FSharpType) (displayContext: FSharpDisplayContext) range pushToHintMode =
+        let displayContext = displayContext.WithShortTypeNames(true)
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:107]: I wonder if we could reuse the display context between symbols. We use short names anyway, so the context is not needed (much?), as the main thing it brings here is what namespaces are open.

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+        TypeHintHighlighting(fcsType.Format(displayContext), range, pushToHintMode)
+
+    let getReturnTypeHint (decl: IParameterOwnerMemberDeclaration) pushToHintMode =
+        match decl with
+        | :? IConstructorDeclaration
+        | :? IAccessorDeclaration -> ValueNone
+        | _ ->
+
+        let equalsToken = decl.EqualsToken
+        let patterns = decl.ParameterPatterns
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:117]: Do we really need to enumerate all the patterns and store them in a collection? Can we just check that parameter declaration sequence is not empty?

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+                memberDeclaration.NameIdentifier.GetDocumentRange().EndOffsetRange()
+
+            | _ -> equalsToken.GetDocumentRange().StartOffsetRange()
+
+        let symbolUse =
+            match decl with
+            | :? IBinding as binding ->
+                let refPat = binding.HeadPattern.As<IReferencePat>()
+                if isNull refPat then Unchecked.defaultof<_> else
+                refPat.GetFcsSymbolUse()
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:134]: Nit: a `match` expression could probably look cleaner.

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+        | :? IParametersOwnerPat as pattern ->
+            let asPat = AsPatNavigator.GetByLeftPattern(pattern.IgnoreParentParens())
+            if isNull asPat then ValueNone else
+
+            let reference = pattern.Reference
+            if isNull reference then ValueNone else
+
+            let symbol = reference.GetFcsSymbol().As<FSharpActivePatternCase>()
+            if isNull symbol then ValueNone
+            else getHintForPattern (asPat.RightPattern.IgnoreInnerParens()) pushToHintMode
```
Comment (auduchinok) [inline ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:158]: Same thing about a possible `match` expression.

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs (hunk 9 above)
Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:207]: I'm fine with merging it as is for now, but we should really think about rewriting this and that pieces of code.

**Review by auduchinok — APPROVED** [review]

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/FSharp.Psi.Daemon.fsproj (hunk 13 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/FSharp.Psi.Daemon.fsproj:17]: I touched fsproj around this place and decided to bring the position of neighboring files to a more expected state. We can also do this in a separate PR.

File: ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs (hunk 14 above)
Comment (DedSec256) [reply ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:107]: Maybe we could just use a default context with short names?

Comment (auduchinok) [reply ReSharper.FSharp/src/FSharp/FSharp.Psi.Daemon/src/Stages/PatternTypeHintsStage.fs:107]: If there's one, and it works with it, then, for sure, let's do it.
